nativelib/src/main/resources/gc/commix/Allocator.c

```diff
@@ -0,0 +1,427 @@
+        fflush(stdout);
+#endif
+        if (block == NULL) {
+            return false;
+        }
+        blockStart = BlockMeta_GetBlockStart(blockMetaStart,
+                                             allocator->heapStart, block);
+
+        allocator->cursor = blockStart;
+        allocator->limit = Block_GetBlockEnd(blockStart);
+        BlockMeta_SetFirstFreeLine(block, LAST_HOLE);
+    }
+
+    allocator->block = block;
+    allocator->blockStart = blockStart;
+
+    return true;
+}
+
+INLINE
+word_t *Allocator_lazySweep(Heap *heap, uint32_t size) {
+    word_t *object = NULL;
+    Stats_DefineOrNothing(stats, heap->stats);
+    Stats_RecordTime(stats, start_ns);
+    // mark as active
+    heap->lazySweep.lastActivity = BlockRange_Pack(1, heap->sweep.cursor);
+    while (object == NULL && heap->sweep.cursor < heap->sweep.limit) {
+        Sweeper_Sweep(heap, heap->stats, &heap->lazySweep.cursorDone,
+                      LAZY_SWEEP_MIN_BATCH);
+        object = Allocator_tryAlloc(&allocator, size);
+    }
+    // mark as inactive
+    heap->lazySweep.lastActivity = BlockRange_Pack(0, heap->sweep.cursor);
+    while (object == NULL && !Sweeper_IsSweepDone(heap)) {
+        object = Allocator_tryAlloc(&allocator, size);
+        if (object == NULL) {
+            sched_yield();
+        }
+    }
+    Stats_RecordTime(stats, end_ns);
+    Stats_RecordEvent(stats, event_sweep, start_ns, end_ns);
+    return object;
+}
+
+NOINLINE word_t *Allocator_allocSlow(Heap *heap, uint32_t size) {
+    word_t *object = Allocator_tryAlloc(&allocator, size);
+
+    if (object != NULL) {
+    done:
+        assert(Heap_IsWordInHeap(heap, object));
+        assert(object != NULL);
+        ObjectMeta *objectMeta = Bytemap_Get(allocator.bytemap, object);
+#ifdef DEBUG_ASSERT
+        ObjectMeta_AssertIsValidAllocation(objectMeta, size);
+#endif
+        ObjectMeta_SetAllocated(objectMeta);
+        return object;
+    }
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Allocator_lazySweep(heap, size);
+
+        if (object != NULL)
+            goto done;
+    }
+
+    Heap_Collect(heap);
+    object = Allocator_tryAlloc(&allocator, size);
+
+    if (object != NULL)
+        goto done;
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Allocator_lazySweep(heap, size);
+
+        if (object != NULL)
+            goto done;
+    }
+
+    // A small object can always fit in a single free block
+    // because it is no larger than 8K while the block is 32K.
+    Heap_Grow(heap, 1);
+    object = Allocator_tryAlloc(&allocator, size);
+
+    goto done;
+}
+
+INLINE word_t *Allocator_Alloc(Heap *heap, uint32_t size) {
+    assert(size % ALLOCATION_ALIGNMENT == 0);
+    assert(size < MIN_BLOCK_SIZE);
+
+    word_t *start = allocator.cursor;
+    word_t *end = (word_t *)((uint8_t *)start + size);
+
+    // Checks if the end of the block overlaps with the limit
+    if (end >= allocator.limit) {
```

Comment:
Why is it `>=` here and not `>` ?

Reply:
You are right, nice catch! I updated that.